packages/wonder-blocks-form/src/components/labeled-text-field.js

```diff
@@ -42,25 +42,28 @@ type Props = {|
     disabled: boolean,
 
     /**
-     * Whether this field is required to to continue.
+     * Whether this field is required to to continue, or the error message to
+     * render if this field is left blank.
      *
-     * If the `validate` prop is not being passed in in conjuction with
-     * the `required` prop, please use the `requiredErrorMessage` prop
-     * to specify the error message.
-     */
-    required?: boolean,
-
-    /**
-     * The error message that should render if this field is
-     * required and left blank. Note that this message will not be used
-     * if a `validate` prop is passed in.
+     * This can be a boolean or a string.
      *
-     * Implementation note: Added this field so that users can pass in
-     * translated strings.
+     * String:
+     * Please pass in a translated string to use as the error message that will
+     * render if the user leaves this field blank. If this field is required,
+     * and a string is not passed in, a default untranslated string will render
+     * upon error.
+     * Note: The string will not be used if a `validate` prop is passed in.
```

Comment:
Given this info, I wonder if we should consolidate our validation API here.

1. `required` is a type of validation. It's a shortcut for "not empt"
2. The error message is what to show if that `required` validation does not pass
3. `validate` is a custom validator

Rather than the `required` prop and `validate` prop, should we have a single `validation` prop that supports validation rules or something?

Not requesting a change right now, but I think it's worth considering.

Imagine a `validation` prop of type `Validation<T> | $ReadOnlyArray<Validation<T>>` where `Validation` is simply a `validate` callback that takes in a value and returns a `?string`. If the string is null/undefined/empty, it's considered valid. `T` is the type of the thing to be validated, so in this case it would be `string`.

We then have helpers:

1. `<T>runValidations(value: ?T, validations: Validation<T> | $ReadOnlyArray<Validation<T>>): ?string`
    This would run all the validations in order and return the first error string.
2. `Validate.Required<T>(translatedErrorMessage?: string)`
    This would be the `required: true` equivalent. You could just set `validation` prop to `Validate.Required<string>()` and that would return a proper `Validation<string>` function that asserted it wasn't empty, for example. Could also take an error string to return if it was missing, rather than the default.
3. `Validate.Custom<T>(validator: (value: T) => ?string): Validation<T>`
     To define a custom rule.
4. Could be other `Validate.` rules like `Validate.Min(minimum: number, translatedErrorMessage: string)` as we needed

I feel this is scalable and reusable and avoids us having these issues where "prop A is only valid if prop B is not set", etc.

Again, not requesting change here, just something for us to think about.

/cc @jandrade

Reply:
I'm wondering if we should just go back to what I was originally going to do, and just have `required` for the asterisk visual, and if someone want to validate the required field, they can do so using the `validate` prop. I was just trying to save work for people writing required fields in the future, but that's how it was being implemented up until now anyway.

Example: 
https://github.com/Khan/webapp/blob/master/services/static/javascript/auth/components/login-form.jsx#L269

Reply:
@somewhatabstract that's super interesting! Agree with you about proposing that change in the near future.

How do you see the `required` (aka add the `*` label) would fit in that model?

Reply:
Yes, good question. There are a couple of things I could imagine.

1. We just have a `markAsRequired` type prop that adds the indicator
2. We have a `decorator` prop that takes a component to render as the decorator, and we can pass in one that's to indicate required.

I think option 1 is probably the best.